components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/username/NotificationUsernameRecoveryManager.java

```diff
@@ -209,3 +101,138 @@
 //        }
 //    }
 //}
+
+    public boolean recoverUserByClaims(ArrayList<UserClaim> claims)
+            throws IdentityRecoveryException {
+
+        // Check whether username recovery enable
+        boolean isRecoveryUsernameEnable = Utils.getRecoveryConfigs().getRecovery().getUsername().isEnable();
+
+        if (!isRecoveryUsernameEnable) {
+
+            if (log.isDebugEnabled()) {
+                log.debug("Username Recovery is not enabled");
+            }
+            return false;
+            //TODO send exception
+//            throw Utils.handleClientException(
+//                    IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_USERNAME_RECOVERY_NOT_ENABLE, null);
+        }
+
+        if (claims == null || claims.size() < 1) {
+
+            if (log.isDebugEnabled()) {
+                log.debug("No claims are recieved");
+            }
+            return false;
+            //TODO send exception
+//            throw Utils.handleClientException(
+//                    IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_NO_FIELD_FOUND_FOR_USER_RECOVERY, null);
+        }
+
+        List<User> resultedUserList = new ArrayList<>();
+        User user;
+
+        // Need to populate the claim email as the first element in the
+        // passed array.
+        for (int i = 0; i < claims.size(); i++) {
+
+            UserClaim claim = claims.get(i);
+            if (StringUtils.isBlank(claim.getClaimValue())) {
+                continue;
+            }
+            if (claim.getClaimURI() != null && claim.getClaimValue() != null) {
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Searching users for " + claim.getClaimURI() + " with the value :"
+                            + claim.getClaimValue());
+                }
+                List<User> matchedUserList = getUserList(claim.getClaimURI(), claim.getClaimValue());
+
+                if (!matchedUserList.isEmpty()) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Matched userList : " + Arrays.toString(matchedUserList.toArray()));
+                    }
+                    //If more than one user find the first matching user list. Hence need to define unique claims
+                    if (!resultedUserList.isEmpty()) {
+                        List<User> users = new ArrayList<>();
+                        for (User resultedUser : resultedUserList) {
+                            for (User matchedUser : matchedUserList) {
+                                if (resultedUser.getUniqueUserId().equals(matchedUser.getUniqueUserId())) {
```

Comment:
Can resultedUser.getUniqueUserId() be null?
If so better to use the null safe StringUtils.equals() method